bundle/bundle.go

```diff
@@ -0,0 +1,398 @@
+			return false, err
+		}
+		if !valid {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// ValidBundle checks if a bundle is syntactically valid.
+// Validates signatures and overall structure.
+func ValidBundle(bundle Bundle) error {
+	var totalSum int64
+
+	sigs := make(map[Hash][]Trytes)
+	k := kerl.NewKerl()
+
+	lastIndex := uint64(len(bundle) - 1)
+	for i := range bundle {
+		tx := &bundle[i]
+		totalSum += tx.Value
+
+		if tx.CurrentIndex != uint64(i) {
+			return errors.Wrapf(ErrInvalidBundle, "expected tx at index %d to have current index %d but got %d", i, i, tx.CurrentIndex)
+		}
+		if tx.LastIndex != lastIndex {
+			return errors.Wrapf(ErrInvalidBundle, "expected tx at index %d to have last index %d but got %d", i, lastIndex, tx.LastIndex)
+		}
+
+		txTrits := MustTrytesToTrits(transaction.MustTransactionToTrytes(tx)[2187 : 2187+162])
+		k.Absorb(txTrits)
+
+		// continue if output or signature tx
+		if tx.Value >= 0 {
+			continue
+		}
+
+		// here we have an input transaction (negative value)
+		sigs[tx.Address] = append(sigs[tx.Address], tx.SignatureMessageFragment)
+
+		// find the subsequent txs containing the remaining signature
+		// message fragments for this input transaction
+		for j := i; j < len(bundle)-1; j++ {
+			tx := &bundle[j+1]
+
+			// check if the tx is part of the input transaction
+			if tx.Address == tx.Address && tx.Value == 0 {
+				// append the signature message fragment
+				sigs[tx.Address] = append(sigs[tx.Address], tx.SignatureMessageFragment)
+			}
+		}
+	}
+
+	// sum of all transaction must be 0
+	if totalSum != 0 {
+		return errors.Wrapf(ErrInvalidBundle, "bundle total sum should be 0 but got %d", totalSum)
+	}
+
+	bundleHashTrits, err := k.Squeeze(HashTrinarySize)
+	if err != nil {
+		return err
+	}
+
+	bundleHash := MustTritsToTrytes(bundleHashTrits)
+
+	if bundleHash != bundle[0].Bundle {
+		return ErrInvalidBundleHash
+	}
+
+	// validate the signatures
+	valid, err := ValidateBundleSignatures(bundle)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return ErrInvalidSignature
+	}
+
+	return nil
+}
+
+// GroupTransactionsIntoBundles groups the given transactions into groups of bundles.
+// Note that the same bundle can exist in the return slice multiple times, though they
+// are reattachments of the same transfer.
+func GroupTransactionsIntoBundles(txs transaction.Transactions) Bundles {
+	bundles := Bundles{}
+
+	for i := range txs {
+		tx := &txs[i]
+		if tx.CurrentIndex != 0 {
+			continue
+		}
+
+		bundle := Bundle{*tx}
+		lastIndex := int(tx.LastIndex)
+		current := tx
+		for x := 1; x <= lastIndex; x++ {
+			// get all txs belonging into this bundle
+			found := false
+			for j := range txs {
+				if current.Bundle != txs[j].Bundle ||
+					txs[j].CurrentIndex != current.CurrentIndex+1 ||
+					current.TrunkTransaction != txs[j].Hash {
+					continue
+				}
+				found = true
+				bundle = append(bundle, txs[j])
+				current = &txs[j]
+				break
+			}
+			if !found {
+				break
```

Comment:
"tail transaction" might have a wrong last index but is still propagated through the network, thereby simply the bundle is returned with only the tail transaction in it. If the function would return an error in case of this invalid/incomplete bundle, then `GetBundlesFromAddresses` would have to return an error too for no good reason. Library user should check him/herself whether the bundles are valid. I've added a note that the function doesn't validate the "constructed" bundles.

Reply:
Fair enough, I do agree that if they ask for a bundle, they should get the bundle, however the validity